crypto/rand/rand_unix.c

```diff
@@ -104,203 +94,103 @@ int RAND_poll(void)
-                     */
-                    try_read = 1;
-                } else {
-                    FD_ZERO(&fset);
-                    FD_SET(fd, &fset);
-
-                    if (select(fd + 1, &fset, NULL, NULL, &t) >= 0) {
-                        usec = t.tv_usec;
-                        if (FD_ISSET(fd, &fset))
-                            try_read = 1;
-                    } else
-                        usec = 0;
-                }
+#    error "OS seeding requires DEVRANDOM to be configured"
 #   endif
+#  endif
+
+int RAND_poll(void)
+{
+#  ifdef OPENSSL_RAND_SEED_NONE
+    return 0;
+#  else
+    int ok = 0;
+    char temp[RANDOMNESS_NEEDED];
+#   define TEMPSIZE (int)sizeof(temp)
+
+# ifdef OPENSSL_RAND_SEED_RDTSC
+    {
+        unsigned char c;
+        int i;
+
+        for (i = 0; i < 10; i++) {
+            /* Low-order bits of the high-speed clock. */
+            c = (unsigned char)(OPENSSL_rdtsc() & 0xFF);
+            RAND_add(&c, 1, 0.5);
+        }
+    }
+# endif
 
-                if (try_read) {
-                    r = read(fd, (unsigned char *)tmpbuf + n,
-                             RANDOMNESS_NEEDED - n);
-                    if (r > 0)
-                        n += r;
-                } else
-                    r = -1;
+#   ifdef OPENSSL_RAND_SEED_EGD
+    {
+        static const char *paths[] = { DEVRANDOM_EGD, NULL };
+        int i;
 
-                /*
-                 * Some Unixen will update t in select(), some won't.  For
-                 * those who won't, or if we didn't use select() in the first
-                 * place, give up here, otherwise, we will do this once again
-                 * for the remaining time.
-                 */
-                if (usec == 10 * 1000)
-                    usec = 0;
+        for (i = 0; paths[i] != NULL; i++) {
+            if (RAND_query_egd_bytes(paths[i], temp, TEMPSIZE) == TEMPSIZE) {
+                RAND_add(temp, TEMPSIZE, TEMPSIZE);
+                ok++;
             }
-            while ((r > 0 ||
-                    (errno == EINTR || errno == EAGAIN)) && usec != 0
-                   && n < RANDOMNESS_NEEDED);
-
-            close(fd);
         }
     }
-#  endif                        /* defined(DEVRANDOM) */
-
-#  if !defined(OPENSSL_NO_EGD) && defined(DEVRANDOM_EGD)
-    /*
-     * Use an EGD socket to read randomness from the daemon.
-     */
+#   endif
 
-    for (egdsocket = egdsockets; *egdsocket && n < RANDOMNESS_NEEDED;
-         egdsocket++) {
-        int r;
+#   ifdef OPENSSL_RAND_SEED_DEVRANDOM
+    {
+        static const char *paths[] = { DEVRANDOM, NULL };
+        FILE *fp;
+        int i;
 
-        r = RAND_query_egd_bytes(*egdsocket, (unsigned char *)tmpbuf + n,
-                                 RANDOMNESS_NEEDED - n);
-        if (r > 0)
-            n += r;
+        for (i = 0; paths[i] != NULL; i++) {
+            if ((fp = fopen(paths[i], "r")) == NULL)
```

Comment:
On side note one can also wonder why does it have to be fopen. Normally one chooses fopen, because it's considered more portable, since it's in *language* specification. But this module is not supposed to be *that* portable.